alternate_test.go

```diff
@@ -52,76 +71,173 @@ func stubNextHandler(rcode int, err error) test.Handler {
+				testNum, rcode, err)
+		}
+
+		// Ensure that overall returned code is correct
+		if rec.Rcode != tc.expectedRcode {
+			t.Errorf("Test '%d': Alternate returned code '%v (%d)', but expected '%v (%d)'",
+				testNum, dns.RcodeToString[rec.Rcode], rec.Rcode, dns.RcodeToString[tc.expectedRcode], tc.expectedRcode)
+		}
+
+		// Ensure that server was called required number of times
+		if srv.called != tc.called {
+			t.Errorf("Test '%d': Server expected to be called %d time(s) but called %d times(s)",
+				testNum, tc.called, srv.called)
+		}
+	}
+}
+
+func TestAlternateMultipleCalls(t *testing.T) {
+	// dummy Upstreams for servicing a specific rcode
+	srv := newServer(dns.RcodeRefused)
+	defer srv.Close()
+	testRules := map[int]rule{
+		dns.RcodeNXRrset:       {forward: srv.f},
+		dns.RcodeServerFailure: {forward: srv.f},
+	}
+
+	testCases := []struct {
+		nextRcode int
+		called    int
+	}{
+		{nextRcode: dns.RcodeNXRrset, called: 10},
+		// No RcodeBadName in table. So, no calls to test server made.
+		{nextRcode: dns.RcodeBadName, called: 0},
+	}
+
+	for testNum, tc := range testCases {
+		srv.Reset()
+		handler := New()
+		// create stub handler to return the test rcode
+		handler.Next = stubNextHandler(tc.nextRcode, nil)
+		// add rules
+		handler.rules = testRules
+
+		// Prepare query and make 10 calls
+		for i := 0; i < 10; i++ {
+			makeTestCall(handler)
+		}
+
+		// Ensure that server was called required number of times
+		if srv.called != tc.called {
+			t.Errorf("Test '%d': Server expected to be called %d time(s) but called %d times(s)",
+				testNum, tc.called, srv.called)
+		}
+	}
+}
+
+func TestAlternateOriginal(t *testing.T) {
+	// dummy Upstreams for servicing a specific rcode
+	srv := newServer(dns.RcodeRefused)
+	defer srv.Close()
+	testRules := map[int]rule{
+		dns.RcodeNXRrset:       {original: true, forward: srv.f},
+		dns.RcodeServerFailure: {forward: srv.f},
+	}
+
+	testCases := []struct {
+		nextRcode int
+		isEdns0   bool
+	}{
+		// isEdns0 is rewrited by original
+		{nextRcode: dns.RcodeNXRrset, isEdns0: false},
+		// RcodeServerFailure hasn't original flag set. isEdns0 remains the same
+		{nextRcode: dns.RcodeServerFailure, isEdns0: true},
+	}
+
+	for testNum, tc := range testCases {
+		srv.Reset()
+		handler := New()
+		// One of rules has "original" flag set
+		handler.original = true
+		// create stub handler to return the test rcode
+		handler.Next = stubNextHandler(tc.nextRcode, nil)
+		// add rules
+		handler.rules = testRules
+
+		// Prepare query and make a call
+		makeTestCall(handler)
+
+		// Ensure edns0 option has expected state
+		if tc.isEdns0 && srv.lastIsEdns0 != tc.isEdns0 {
+			t.Errorf("Test '%d': Server expected to recieve Edns0, but didn't", testNum)
+		}
+		if !tc.isEdns0 && srv.lastIsEdns0 != tc.isEdns0 {
+			t.Errorf("Test '%d': Server expected to recieve no Edns0, but received it",
+				testNum)
+		}
```

Comment:
IMHO, the above two tests can be combined to one. All we are interested in is whether `srv.lastIsEdns0 == tc.isEdns0`.